File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/wizard/WizardShell.kt
@file:Suppress("LongMethod")

package app.pantopus.android.ui.screens.shared.wizard

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.components.GhostButton
import app.pantopus.android.ui.components.PrimaryButton
import app.pantopus.android.ui.components.SegmentedProgressBar
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Spacing

/** Test-tag constants for the wizard chrome. */
object WizardShellTags {
    const val SHELL = "wizardShell"
    const val LEADING = "wizardLeadingButton"
    const val PRIMARY_CTA = "wizardPrimaryCTA"
    const val STEP_READOUT = "wizardStepReadout"
}

/**
 * Generic wizard chrome — top bar (X/back + title + N/M readout),
 * segmented progress bar, scrolling content, and a sticky bottom CTA row
 * with a primary button + optional ghost. Owns the discard-confirm dialog.
 *
 * Concrete wizards supply a [WizardModel] and a content composable; the
 * shell handles everything else.
 */
@Composable
fun WizardShell(
    model: WizardModel,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    var showDiscard by remember { mutableStateOf(false) }
    val chrome = model.chrome

    Scaffold(
        modifier = modifier.fillMaxSize().testTag(WizardShellTags.SHELL),
        containerColor = PantopusColors.appBg,
        topBar = {
            Column(modifier = Modifier.fillMaxWidth().background(PantopusColors.appSurface)) {
                WizardTopBar(
                    title = chrome.title,
                    leading = chrome.leading,
                    progressLabel = chrome.progressLabel,
                    onLeading = {
                        if (chrome.leading == WizardLeadingControl.Close && chrome.dirty) {
                            showDiscard = true
                        } else {
                            model.onLeading()
                        }
                    },
                )
                if (chrome.showsProgressBar) {
                    val total =
                        when (val label = chrome.progressLabel) {
                            is WizardProgressLabel.StepOf -> label.total
                            WizardProgressLabel.Hidden -> 1
                        }
                    val filled =
                        chrome.progressFraction
                            ?.coerceIn(0f, 1f)
                            ?.let { (it * total).toInt() }
                            ?: 0
                    SegmentedProgressBar(
                        filled = filled,
                        total = total.coerceAtLeast(1),
                        modifier = Modifier
                            .padding(horizontal = Spacing.s4, vertical = Spacing.s2),
                    )
                }
            }
        },
        bottomBar = {
            WizardStickyCta(
                chrome = chrome,
                onPrimary = model::onPrimary,
                onSecondary = model::onSecondary,
            )
        },
    ) { padding: PaddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .background(PantopusColors.appBg),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = Spacing.s4, vertical = Spacing.s4),
                verticalArrangement = Arrangement.spacedBy(Spacing.s5),
            ) {
                content()
            }
        }
    }

    WizardCloseConfirm(
        visible = showDiscard,
        onDiscard = {
            showDiscard = false
            model.onDiscard()
        },
        onKeepGoing = { showDiscard = false },
    )
}

@Composable
private fun WizardTopBar(
    title: String,
    leading: WizardLeadingControl,
    progressLabel: WizardProgressLabel,
    onLeading: () -> Unit,
) {
    val interaction = remember { MutableInteractionSource() }
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(min = 44.dp)
            .background(PantopusColors.appSurface)
            .padding(horizontal = Spacing.s2),
    ) {
        Text(
            text = title,
            style = PantopusTextStyle.body,
            color = PantopusColors.appText,
            modifier = Modifier
                .align(Alignment.Center)
                .semantics { heading() },
        )
        Box(
            modifier = Modifier
                .align(Alignment.CenterStart)
                .size(44.dp)
                .testTag(WizardShellTags.LEADING)
                .semantics {
                    contentDescription = if (leading == WizardLeadingControl.Close) "Close" else "Back"
                    role = Role.Button
                }
                .clickable(
                    interactionSource = interaction,
                    indication = null,
                    onClick = onLeading,
                ),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(
                icon = if (leading == WizardLeadingControl.Close) PantopusIcon.X else PantopusIcon.ChevronLeft,
                contentDescription = null,
                size = 22.dp,
            )
        }
        if (progressLabel is WizardProgressLabel.StepOf) {
            Text(
                text = "${progressLabel.current} of ${progressLabel.total}",
                style = PantopusTextStyle.caption,
                color = PantopusColors.appTextSecondary,
                modifier = Modifier
                    .align(Alignment.CenterEnd)
                    .padding(end = Spacing.s3)
                    .testTag(WizardShellTags.STEP_READOUT),
            )
        }
    }
    HorizontalDivider(thickness = 1.dp, color = PantopusColors.appBorderSubtle)
}

@Composable
private fun WizardStickyCta(
    chrome: WizardChrome,
    onPrimary: () -> Unit,
    onSecondary: () -> Unit,
) {
    Column(modifier = Modifier.fillMaxWidth()) {
        HorizontalDivider(thickness = 1.dp, color = PantopusColors.appBorderSubtle)
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(PantopusColors.appSurface)
                .padding(Spacing.s4),
            horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
        ) {
            chrome.secondaryCta?.let { secondary ->
                GhostButton(
                    title = secondary.label,
                    onClick = onSecondary,
                    modifier = Modifier.weight(1f).testTag(secondary.testTag),
                )
            }
            PrimaryButton(
                title = if (chrome.isSubmitting) "Working…" else chrome.primaryCtaLabel,
                onClick = onPrimary,
                isLoading = chrome.isSubmitting,
                isEnabled = chrome.primaryCtaEnabled,
                modifier = Modifier.weight(1f).testTag(WizardShellTags.PRIMARY_CTA),
            )
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Shared/Wizard/WizardShell.swift
//
//  WizardShell.swift
//  Pantopus
//
//  Generic wizard chrome: top bar (X/back + title + N/M readout), segmented
//  progress, scrolling content, and a sticky bottom CTA row with primary +
//  optional ghost. Owns the discard-confirm sheet.
//

import SwiftUI

/// Reusable wizard scaffold. Feature screens supply a `WizardModel` and a
/// content view; everything else (chrome, progress, CTA, close-confirm)
/// is handled here.
///
/// Pass `identity` to repaint the progress rail and the primary CTA in
/// a non-default pillar (`.home`, `.business`, `.warm`). Default is
/// `.personal` so legacy call sites render identically.
@MainActor
public struct WizardShell<Content: View>: View {
    private let model: any WizardModel
    private let identity: WizardIdentity
    private let content: Content

    @State private var showsDiscard = false

    public init(
        model: any WizardModel,
        identity: WizardIdentity = .personal,
        @ViewBuilder content: () -> Content
    ) {
        self.model = model
        self.identity = identity
        self.content = content()
    }

    public var body: some View {
        let chrome = model.chrome
        return VStack(spacing: Spacing.s0) {
            WizardTopBar(
                title: chrome.title,
                leading: chrome.leading,
                progressLabel: chrome.progressLabel,
                onLeading: handleLeading
            )
            if chrome.showsProgressBar, let fraction = chrome.progressFraction {
                progressBar(fraction: fraction)
            }
            ScrollView {
                VStack(alignment: .leading, spacing: Spacing.s5) {
                    content
                }
                .padding(.horizontal, Spacing.s4)
                .padding(.top, Spacing.s4)
                .padding(.bottom, Spacing.s8)
            }
            .background(Theme.Color.appBg)
            stickyCTA(chrome: chrome)
        }
        .background(Theme.Color.appBg)
        .accessibilityIdentifier("wizardShell")
        .wizardCloseConfirm(isPresented: $showsDiscard) { model.discardConfirmed() }
    }

    @ViewBuilder
    private func progressBar(fraction: Double) -> some View {
        let segmentCount: Int = if case let .stepOf(_, total) = model.chrome.progressLabel {
            max(1, total)
        } else {
            1
        }
        let filled = Int((Double(segmentCount) * fraction).rounded())
        SegmentedProgressBar(
            currentStep: filled,
            totalSteps: segmentCount,
            fillColor: identity.accent
        )
            .padding(.horizontal, Spacing.s4)
            .padding(.vertical, Spacing.s2)
            .background(Theme.Color.appSurface)
    }

    private func stickyCTA(chrome: WizardChrome) -> some View {
        VStack(spacing: Spacing.s0) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
            HStack(spacing: Spacing.s3) {
                if let secondary = chrome.secondaryCTA {
                    GhostButton(title: secondary.label) {
                        await MainActor.run { model.secondaryTapped() }
                    }
                    .accessibilityIdentifier(secondary.identifier)
                }
                WizardPrimaryCTA(
                    title: chrome.isSubmitting ? "Working…" : chrome.primaryCTALabel,
                    isLoading: chrome.isSubmitting,
                    isEnabled: chrome.primaryCTAEnabled,
                    tint: identity.accent,
                    shadow: identity.ctaShadow
                ) {
                    await MainActor.run { model.primaryTapped() }
                }
                .accessibilityIdentifier("wizardPrimaryCTA")
            }
            .padding(Spacing.s4)
            .background(Theme.Color.appSurface)
        }
    }

    private func handleLeading() {
        switch model.chrome.leading {
        case .back:
            model.leadingTapped()
        case .close:
            if model.chrome.dirty {
                showsDiscard = true
            } else {
                model.leadingTapped()
            }
        }
    }
}

/// 44pt top bar: X/back leading, centered title, "N of M" trailing.
private struct WizardTopBar: View {
    let title: String
    let leading: WizardLeadingControl
    let progressLabel: WizardProgressLabel
    let onLeading: () -> Void

    var body: some View {
        ZStack {
            Text(title)
                .pantopusTextStyle(.body)
                .foregroundStyle(Theme.Color.appText)
                .accessibilityAddTraits(.isHeader)
            HStack {
                Button(action: onLeading) {
                    Icon(leadingIcon, size: 22, color: Theme.Color.appText)
                        .frame(width: 44, height: 44)
                }
                .accessibilityLabel(leadingA11y)
                .accessibilityIdentifier("wizardLeadingButton")
                Spacer()
                Group {
                    if case let .stepOf(current, total) = progressLabel {
                        Text("\(current) of \(total)")
                            .pantopusTextStyle(.caption)
                            .foregroundStyle(Theme.Color.appTextSecondary)
                            .padding(.horizontal, Spacing.s3)
                            .accessibilityIdentifier("wizardStepReadout")
                    } else {
                        Color.clear.frame(width: 44, height: 44)
                    }
                }
            }
            .padding(.horizontal, Spacing.s2)
        }
        .frame(height: 44)
        .background(Theme.Color.appSurface)
        .overlay(alignment: .bottom) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
    }

    private var leadingIcon: PantopusIcon {
        switch leading {
        case .close: .x
        case .back: .chevronLeft
        }
    }

    private var leadingA11y: String {
        switch leading {
        case .close: "Close"
        case .back: "Back"
        }
    }
}

/// Primary CTA used by the wizard sticky row. Mirrors the geometry and
/// loading/disabled behaviour of `PrimaryButton` but takes an explicit
/// `tint` + `shadow` so the wizard can paint the button in the current
/// identity pillar (sky / home / business / warm-amber).
@MainActor
private struct WizardPrimaryCTA: View {
    let title: String
    let isLoading: Bool
    let isEnabled: Bool
    let tint: Color
    let shadow: PantopusShadow
    let action: () async -> Void

    var body: some View {
        Button {
            Task { await action() }
        } label: {
            ZStack {
                Text(title)
                    .pantopusTextStyle(.body)
                    .opacity(isLoading ? 0 : 1)
                if isLoading {
                    ProgressView().tint(Theme.Color.appTextInverse)
                }
            }
            .foregroundStyle(Theme.Color.appTextInverse)
            .frame(maxWidth: .infinity, minHeight: 44)
            .padding(.horizontal, Spacing.s4)
            .background(tint)
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
            .pantopusShadow(shadow)
            .opacity(isEnabled ? 1 : 0.5)
        }
        .buttonStyle(.plain)
        .disabled(!isEnabled || isLoading)
        .accessibilityLabel(title)
        .accessibilityAddTraits(.isButton)
    }
}
